lumicks/pylake/file_download.py

```diff
@@ -0,0 +1,139 @@
+import os
+import hashlib
+from urllib.parse import urljoin
+import requests
+from tqdm.auto import tqdm
+
+
+__all__ = ["download_from_doi"]
+
+
+def get_url_from_doi(doi):
+    """Obtains a Zenodo record from the DOI (e.g. 10.5281/zenodo.#)"""
+    url = doi if doi.startswith("http") else urljoin("https://doi.org/", doi)
+    response = requests.get(url)
+
+    if not response.ok:
+        raise RuntimeError(f"DOI could not be resolved {response}")
+
+    return response.url
+
+
+def download_record_metadata(record_number):
+    """Download specific Zenodo record metadata"""
+    zenodo_url = "https://zenodo.org/api/records/"
+    response = requests.get(urljoin(zenodo_url, str(record_number)))
+    if response.ok:
+        return response.json()
+    else:
+        raise ValueError(f"Failed to access Zenodo record {response}")
+
+
+def download_file(url, target_path, file_name, show_progress=True, block_size=8192):
+    """Stream a file from a URL while showing progress
+
+    Parameters
+    ----------
+    url : str
+        URL to pull from
+    target_path : str
+        Target path to download to
+    file_name : str
+        Filename to download
+    show_progress : bool
+        Show progress while downloading or not?
+    block_size : int
+        Block size to use when downloading
+    """
+    response = requests.get(url, stream=True)
+    size_bytes = int(response.headers.get("content-length", 0))
+
+    full_path = os.path.join(target_path, file_name)
+    dir_name = os.path.dirname(full_path)
+    if dir_name:
+        os.makedirs(dir_name, exist_ok=True)
```

Comment:
`dir_name` is just the same as `target_path`, no? Also, shouldn't it be `if not dir_name:`?

Reply:
Hmm, I think the confusion here comes from `file_name`  being a bit of a misnomer. `file_name` is actually a relative path provided by Zenodo, for the file currently being downloaded. Therefore, it's important to concatenate them, then re-extract the actual directory with `dirname` (otherwise you get a folder with the filename's name). I've renamed the `file_name` argument to `download_path` to clear up this confusion.

Regarding the if-statement, that is as intended. We only attempt to make folders if it needs to be done. Note that the `makedirs` call has `exist_ok=True`, which means that if the directory already exists, it just silently ignores it.

Reply:
much better now